src/SFML/Window/Unix/WindowImplX11.cpp

```diff
@@ -1061,54 +1063,128 @@ void WindowImplX11::setVideoMode(const VideoMode& mode)
+        {
+            XRRFreeOutputInfo(outputInfo);
+        }
 
-    // Search for a matching size
-    for (int i = 0; (sizes && i < nbSizes); ++i)
+        // Failed to get output info
+        err() << "Failed to get output info for fullscreen mode, switching to window mode" << std::endl;
+        return;
+    }
+
+    // Get crtc info from crtc
+    XRRCrtcInfo* crtcInfo = XRRGetCrtcInfo(m_display, res, outputInfo->crtc);
+    if (!crtcInfo)
     {
-        XRRConfigRotations(config, &currentRotation);
+        // Free screen resources
+        XRRFreeScreenResources(res);
 
-        if (currentRotation == RR_Rotate_90 || currentRotation == RR_Rotate_270)
-            std::swap(sizes[i].height, sizes[i].width);
+        // Free  output info
+        XRRFreeOutputInfo(outputInfo);
 
-        if ((sizes[i].width  == static_cast<int>(mode.width)) && (sizes[i].height == static_cast<int>(mode.height)))
-        {
-            // Switch to fullscreen mode
-            XRRSetScreenConfig(m_display, config, RootWindow(m_display, m_screen), i, currentRotation, CurrentTime);
+        // Failed to get crtc info
+        err() << "Failed to get crtc info for fullscreen mode, switching to window mode" << std::endl;
+        return;
+    }
+
+    // Find RRMode to set 
+    bool modeFound = false;
+    RRMode xRandMode;
 
-            // Set "this" as the current fullscreen window
-            fullscreenWindow = this;
+    for (int i = 0; i < res->nmode; i++)
+    {
+        if (crtcInfo->rotation == RR_Rotate_90 || crtcInfo->rotation == RR_Rotate_270)
+            std::swap(res->modes[i].height, res->modes[i].width);
+
+        // Check if screen size match
+        if (res->modes[i].width == static_cast<int>(mode.width) && 
+            res->modes[i].height == static_cast<int>(mode.height))
+        {
+            xRandMode = res->modes[i].id;
+            modeFound = true;
             break;
         }
     }
 
-    // Free the configuration instance
-    XRRFreeScreenConfigInfo(config);
+    if (!modeFound)
+    {
+        // Failed to find a matching RRMode
+        err() << "Failed to find a matching RRMode for fullscreen mode, switching to window mode" << std::endl;
+        return;
+    }
+
+    // Save the current video mode before we switch to fullscreen
+    m_oldVideoMode = crtcInfo->mode;
+    m_oldRRCrtc = outputInfo->crtc;
+
+    // Switch to fullscreen mode
+    XRRSetCrtcConfig(m_display,
+                     res,
+                     outputInfo->crtc,
+                     CurrentTime,
+                     crtcInfo->x,
+                     crtcInfo->y,
+                     xRandMode,
+                     crtcInfo->rotation,
+                     &output,
+                     1);
+
+    // Set "this" as the current fullscreen window
+    fullscreenWindow = this;
+
+    // Free screen resources
+    XRRFreeScreenResources(res);
+
+    // Free  output info
+    XRRFreeOutputInfo(outputInfo);
+
+    // Free crtc info
```

Comment:
I think the `// Free <stuff>` comment should be removed -- they don't bring much.